src/main/scala/sbtassembly/Assembly.scala

```diff
@@ -287,136 +287,134 @@ object Assembly {
           ),
         log
       )
-    val (jarFiles, jarFileEntries) = timed(Level.Debug, "Collect and shade dependency entries") {
+    val jarFileEntries = timed(Level.Debug, "Collect and shade dependency entries") {
       filteredJars.par.map { jar =>
         val module = jar.metadata
           .get(moduleID.key)
           .map(m => ModuleCoordinate(m.organization, m.name, m.revision))
           .getOrElse(ModuleCoordinate("", jar.data.name.replaceAll(".jar", ""), ""))
         val jarFile = new JarFile(jar.data)
-        jarFile -> jarFile
+        val jarFileEntriesRes = jarFile
           .entries()
           .asScala
           .filterNot(_.isDirectory)
           .toVector
           .par
           .flatMap { entry =>
-            jarShader(module)(entry.getName, () => jarFile.getInputStream(entry))
+            jarShader(module)(entry.getName, () => new ByteArrayInputStream(jarFile.getInputStream(entry).readAllBytes()))
```

Comment:
I don't know if this is a good idea. Maybe it would be better to create a setting to tweak the parallelism of `.par` to reduce the concurrent calls on very big JARs?

Reply:
I guess you're right. The PR as it stands requires heap equivalent to all JAR file size right?

Reply:
Jar Jar Abrams needs to read the streams to process various rules against the stream of bytecode I think. Similar to what it's doing `() => jarFile.getInputStream(entry)` a lazy function that creates a stream on demand, I guess we can further make it lazier to provide:

```scala
name => {
  val jarFile = lookupJarFile(jar.data)
  val entry = jarFile.getEntry(name)
  jarFile.getInputStream(entry)
}
```

where we can keep a LRU cache of JAR files in `lookupJarFile(...)`?

Reply:
The shader that rely on this jarjarabrams Shader seems to already load the whole stream into a [ByteArrayInputStream](https://github.com/sbt/sbt-assembly/blob/develop/src/main/scala/sbtassembly/Assembly.scala#L503)
From jarjar-abrams bytecodeShader signature: https://github.com/eed3si9n/jarjar-abrams/blob/develop/core/src/main/scala/com/eed3si9n/jarjarabrams/Shader.scala#L71 

I'm not sure to see the gain in keeping the stream as we will load all the bytes in the shader. Also wondering if we rely on such mechanism with LRU cache if we could not reach some bugs with LRU cache closing the file handler while still having an input stream open from it